### PR TITLE
Mpoly refactor

### PR DESCRIPTION
**Pull request by srajangarg**
WIP

### CODE REVIEW

**Comment by certik** [comment]
@srajangarg I am not up to speed with the recent polys development. Is the `Poly` class templated on a particular polynomial, and it allows you to use any polynomial inside a SymEngine expression?

I forgot how the Piranha and Flint polynomials fit in.

Do you think you could please document how polynomials work in symengine, probably as a markdown document in the doc directory? So that new people can get up to speed how things work and what the motivation for each class is (even if it's not fully implemented yet).

The PR looks ok to me, but I'll let @isuruf review it.

**Comment by srajangarg** [comment]
@certik Sure I will do that. This PR is a WIP btw, more work to follow.

**Comment by srajangarg** [comment]
I'm leaving this for now, please review @isuruf 
I am working on conversion now.

File: symengine/dict.h [hunk 1]
```diff
@@ -49,20 +49,30 @@ typedef std::map<unsigned, integer_class> map_uint_mpz;
...
+class vec_hash
+{
+public:
+    std::size_t operator()(const T &v) const
     {
         std::size_t h = 0;
-        for (const auto &p : k) {
-            h = (h << 4) + p;
+        for (auto i : v) {
+            h ^= i + 0x9e3779b + (h << 6) + (h >> 2);
```
Comment (certik) [inline symengine/dict.h]: This will slow some of the benchmarks.

Why is this new hash function needed?

Comment (srajangarg) [inline symengine/dict.h]: Initially `vector<uint>` used the above hash function, while `vector<int>` used the previous one.
I thought it would make sense to use the same hash functions for both. I also remember Isuru saying that the (older) hash function for `vector<int>` wasn't very good.

Comment (certik) [inline symengine/dict.h]: The older hash function isn't very good, but it's fast. So that's the trade off. Let's keep both, so that we can at least easily switch them.

Comment (isuruf) [inline symengine/dict.h]: It's better to use `hash_combine` to combine hashes instead of writing algorithms like this. Only place where we don't use `hash_combine` is `Add`, but `Add` is special than others. If `hash_combine` isn't good, we can change that later.

File: symengine/polynomial_multivariate.cpp [hunk 2]
```diff
@@ -92,66 +69,42 @@ MultivariateIntPolynomial::convert(const UIntPoly &o)
 unsigned int reconcile(vec_uint &v1, vec_uint &v2, set_basic &s,
                        const set_basic &s1, const set_basic &s2)
 {
-    auto a1 = s1.begin();
-    auto a2 = s2.begin();
-    unsigned int poscount = 0;
-    // Performs a merge sort of s1 and s2, and builds up v1 and v2 as
-    // translators:
-    // v[i] is the position of the ith symbol in the new set.
```
Comment (isuruf) [inline symengine/polynomial_multivariate.cpp]: Can you keep this comment?

File: symengine/polynomial_multivariate.h [hunk 3]
```diff
@@ -22,14 +21,13 @@ unsigned int reconcile(vec_uint &v1, vec_uint &v2, set_basic &s,
                        const set_basic &s1, const set_basic &s2);
 // translates vectors from one polynomial into vectors for another.
 template <typename Vec>
-Vec translate(Vec original, vec_uint translator, unsigned int size)
+void translate(const Vec &original, Vec &changed, vec_uint translator,
```
Comment (isuruf) [inline symengine/polynomial_multivariate.h]: Why is this change needed?

Comment (srajangarg) [inline symengine/polynomial_multivariate.h]: To prevent the return by copy

Comment (isuruf) [inline symengine/polynomial_multivariate.h]: Because of copy elision optimization, it won't.

**Comment by isuruf** [comment]
+1

File: symengine/dict.h (hunk 1 above)
Comment (certik) [inline symengine/dict.h]: @srajangarg, @isuruf I created #1057 to investigate hash functions.

**Comment by isuruf** [comment]
Can you fix the errors?

**Comment by srajangarg** [comment]
I have been working on a new structure all together, will push by the end of the day.

**Comment by srajangarg** [comment]
@isuruf any comments on the new changes?

File: symengine/polynomial_multivariate.cpp [hunk 4]
```diff
@@ -1,220 +0,0 @@
...
-            term *= temp;
-            whichvar++;
-        }
-        ans += term;
-    }
-    return ans;
-}
-
-RCP<const MultivariateIntPolynomial>
-MultivariateIntPolynomial::convert(const UIntPoly &o)
```
Comment (isuruf) [inline symengine/polynomial_multivariate.cpp]: Why remove this function?

File: symengine/printer.cpp [hunk 5]
```diff
@@ -712,7 +712,7 @@ void StrPrinter::bvisit(const MultivariatePolynomial &x)
                     expr << "*";
                 }
                 expr << it->__str__();
-                if (exps[i] > 1)
+                if (exps[i] > 1 or exps[i] < 0)
```
Comment (isuruf) [inline symengine/printer.cpp]: Can you add a test for this?

Comment (srajangarg) [inline symengine/printer.cpp]: Yes, I will. But it's a correction for negative exponents in the Expression polynomials. those don't get printed, in the previous code.

File: symengine/polynomial_multivariate.cpp (hunk 4 above)
Comment (srajangarg) [inline symengine/polynomial_multivariate.cpp]: So, right now, I have removed the functionality for converting univariate polynomials to their multivariate counterparts. Which also made me change the `add_mpoly` etc. functions. Now, all we need is a general  purpose `add_poly` etc. functions, which do the "type" of poly check (multi/uni?) and their implementation (flint/piranha)? . Then the necessary conversions will will be done and the specific `add_mpoly` or `add_upoly` function will be called.

File: symengine/printer.cpp (hunk 5 above)
Comment (isuruf) [inline symengine/printer.cpp]: Yes, we want the correction to be in the tests, so that this bug will not appear again.

File: symengine/polynomial_multivariate.cpp (hunk 4 above)
Comment (isuruf) [inline symengine/polynomial_multivariate.cpp]: How do you plan to do this? Remember you have to consider the type of polys as well (Int, Rat, Expr)

**Comment by srajangarg** [comment]
@isuruf I tried inserting into an ordered map, the contents of the unordered map. Still no dice. The same order is observed in MSCV/MinGW builds.

**Comment by isuruf** [comment]
As I said on gitter, this is because the ordering of the variables in `set_basic` is different in MSVC vs others.

**Comment by srajangarg** [comment]
No, here I am talking about the ordering of the monomials itself (which is defined by the vectors/keys of the map). And not the ordering of variables inside the monomial.

ie. x**2*y + 1 differing from 1 + x**2*y
This is not caused by the ordering in `set_basic`

**Comment by isuruf** [comment]
Hmm, I only see the following failure on appveyor
```
"a*y**-1*x - 3 - b*x"
  ==
  "-3 + a*x*y**-1 - b*x"
```

Can you provide a link to the failure in the log?

**Comment by srajangarg** [comment]
No, that is the error. It is an error in  the ordering of the monomials right?

**Comment by srajangarg** [comment]
@isuruf Can I get this merged in without the printing test? (I'll add an issue) I can then proceed with Basic -> MPoly.

**Comment by isuruf** [comment]
Sure
